File: packages/turbopuffer-fs/src/schema.ts
import { Buffer } from "node:buffer";
import { createHash } from "node:crypto";
import { basename, extension, normalizePath, parentPath, ancestorPaths, pathId } from "./paths.js";
import type { FsRow, FsSchema } from "./types.js";

const TEXT_EXTENSIONS = new Set([
  ".c",
  ".cfg",
  ".conf",
  ".cpp",
  ".css",
  ".csv",
  ".html",
  ".ini",
  ".java",
  ".js",
  ".json",
  ".jsx",
  ".md",
  ".py",
  ".r",
  ".rb",
  ".rs",
  ".rst",
  ".sh",
  ".sql",
  ".svg",
  ".toml",
  ".ts",
  ".tsx",
  ".txt",
  ".tsv",
  ".xml",
  ".yaml",
  ".yml",
]);

export const META_FIELDS = [
  "id",
  "path",
  "parent",
  "basename",
  "kind",
  "ext",
  "mime",
  "size_bytes",
  "is_text",
  "sha256",
  "source_mtime_ns",
  "source_size_bytes",
] as const;

export const CONTENT_FIELDS = [...META_FIELDS, "text", "blob_b64"] as const;

export function fsSchema(): FsSchema {
  return {
    path: { type: "string", regex: true, filterable: true },
    parent: "string",
    basename: { type: "string", regex: true, filterable: true },
    kind: "string",
    ext: "string",
    mime: "string",
    size_bytes: "uint",
    is_text: "uint",
    text: { type: "string", regex: true, filterable: true },
    blob_b64: { type: "string", filterable: false },
    sha256: "string",
    source_mtime_ns: "uint",
    source_size_bytes: "uint",
  };
}

export function metadataRow(row: FsRow): FsRow {
  return Object.fromEntries(
    META_FIELDS.filter((field) => field in row).map((field) => [field, row[field]]),
  ) as FsRow;
}

export function contentRow(row: FsRow): FsRow {
  return Object.fromEntries(
    CONTENT_FIELDS.filter((field) => field in row).map((field) => [field, row[field]]),
  ) as FsRow;
}

export function sha256Hex(data: Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

function inferMime(path: string, fallback: string): string {
  const ext = extension(path).toLowerCase();
  const table: Record<string, string> = {
    ".txt": "text/plain",
    ".md": "text/markdown",
    ".csv": "text/csv",
    ".tsv": "text/tab-separated-values",
    ".json": "application/json",
    ".py": "text/x-python",
    ".js": "text/javascript",
    ".ts": "text/typescript",
    ".tsx": "text/typescript",
    ".html": "text/html",
    ".css": "text/css",
    ".xml": "application/xml",
    ".yaml": "application/yaml",
    ".yml": "application/yaml",
    ".toml": "application/toml",
    ".sql": "application/sql",
    ".svg": "image/svg+xml",
  };
  return table[ext] ?? fallback;
}

function withSourceMetadata(
  row: FsRow,
  sourceMtimeNs?: number,
  sourceSizeBytes?: number,
): FsRow {
  if (sourceMtimeNs !== undefined) {
    row.source_mtime_ns = sourceMtimeNs;
  }
  if (sourceSizeBytes !== undefined) {
    row.source_size_bytes = sourceSizeBytes;
  }
  return row;
}

function baseRow(path: string, params: {
  kind: "file" | "dir";
  mime: string;
  sizeBytes: number;
  isText: 0 | 1;
  sha256?: string;
}): FsRow {
  const normalized = normalizePath(path);
  const row: FsRow = {
    id: pathId(normalized),
    path: normalized,
    basename: basename(normalized),
    kind: params.kind,
    ext: params.kind === "dir" ? "" : extension(normalized),
    mime: params.mime,
    size_bytes: params.sizeBytes,
    is_text: params.isText,
  };
  const parent = parentPath(normalized);
  if (parent !== undefined) {
    row.parent = parent;
  }
  if (params.sha256 !== undefined) {
    row.sha256 = params.sha256;
  }
  return row;
}

export function directoryRow(path: string, sourceMtimeNs?: number): FsRow {
  return withSourceMetadata(
    baseRow(path, {
      kind: "dir",
      mime: "inode/directory",
      sizeBytes: 0,
      isText: 0,
    }),
    sourceMtimeNs,
    0,
  );
}

export function textRow(
  path: string,
  text: string,
  mime?: string,
  sourceMtimeNs?: number,
  sourceSizeBytes?: number,
): FsRow {
  const data = new TextEncoder().encode(text);
  const row = baseRow(path, {
    kind: "file",
    mime: mime ?? inferMime(path, "text/plain"),
    sizeBytes: data.length,
    isText: 1,
    sha256: sha256Hex(data),
  });
  row.text = text;
  return withSourceMetadata(row, sourceMtimeNs, sourceSizeBytes ?? data.length);
}

export function bytesRow(
  path: string,
  data: Uint8Array,
  mime?: string,
  sourceMtimeNs?: number,
  sourceSizeBytes?: number,
): FsRow {
  const row = baseRow(path, {
    kind: "file",
    mime: mime ?? inferMime(path, "application/octet-stream"),
    sizeBytes: data.length,
    isText: 0,
    sha256: sha256Hex(data),
  });
  row.blob_b64 = Buffer.from(data).toString("base64");
  return withSourceMetadata(row, sourceMtimeNs, sourceSizeBytes ?? data.length);
}

export function parentDirectoryRows(path: string): FsRow[] {
  return ancestorPaths(path, false).map((ancestor) => directoryRow(ancestor));
}

export function targetDirectoryRows(path: string): FsRow[] {
  return ancestorPaths(path, true).map((ancestor) => directoryRow(ancestor));
}

export function isProbablyText(path: string, data: Uint8Array): boolean {
  const ext = extension(path).toLowerCase();
  if (data.includes(0)) {
    return false;
  }
  if (TEXT_EXTENSIONS.has(ext)) {
    return true;
  }
  const mime = inferMime(path, "application/octet-stream");
  if (
    mime.startsWith("text/") ||
    ["application/json", "application/sql", "application/xml", "application/x-sh", "image/svg+xml"].includes(
      mime,
    )
  ) {
    return true;
  }
  try {
    new TextDecoder("utf-8", { fatal: true }).decode(data);
    return true;
  } catch {
    return false;
  }
}

export function rowFromBytes(
  path: string,
  data: Uint8Array,
  mime?: string,
  sourceMtimeNs?: number,
  sourceSizeBytes?: number,
): FsRow {
  if (isProbablyText(path, data)) {
    try {
      const text = new TextDecoder("utf-8", { fatal: true }).decode(data);
      return textRow(path, text, mime, sourceMtimeNs, sourceSizeBytes);
    } catch {
      return bytesRow(path, data, mime, sourceMtimeNs, sourceSizeBytes);
    }
  }
  return bytesRow(path, data, mime, sourceMtimeNs, sourceSizeBytes);
}

export function upsertRowsPayload(rows: FsRow[]): { upsert_rows: FsRow[]; schema: FsSchema } {
  return {
    upsert_rows: rows,
    schema: fsSchema(),
  };
}

File: src/turbopuffer_fs/schema.py
"""Schema and row builders for the turbopuffer filesystem model."""

from __future__ import annotations

import base64
import hashlib
import mimetypes
from typing import Iterable

from .paths import ancestor_paths, basename, extension, normalize_path, parent_path, path_id


TEXT_EXTENSIONS = {
    ".c",
    ".cfg",
    ".conf",
    ".cpp",
    ".css",
    ".csv",
    ".html",
    ".ini",
    ".java",
    ".js",
    ".json",
    ".jsx",
    ".md",
    ".py",
    ".r",
    ".rb",
    ".rs",
    ".rst",
    ".sh",
    ".sql",
    ".svg",
    ".toml",
    ".ts",
    ".tsx",
    ".txt",
    ".tsv",
    ".xml",
    ".yaml",
    ".yml",
}

META_FIELDS = [
    "id",
    "path",
    "parent",
    "basename",
    "kind",
    "ext",
    "mime",
    "size_bytes",
    "is_text",
    "sha256",
    "source_mtime_ns",
    "source_size_bytes",
]

CONTENT_FIELDS = META_FIELDS + ["text", "blob_b64"]


def fs_schema() -> dict[str, object]:
    return {
        "path": {"type": "string", "regex": True, "filterable": True},
        "parent": "string",
        "basename": {"type": "string", "regex": True, "filterable": True},
        "kind": "string",
        "ext": "string",
        "mime": "string",
        "size_bytes": "uint",
        "is_text": "uint",
        "text": {"type": "string", "regex": True, "filterable": True},
        "blob_b64": {"type": "string", "filterable": False},
        "sha256": "string",
        "source_mtime_ns": "uint",
        "source_size_bytes": "uint",
    }


def metadata_row(row: dict[str, object]) -> dict[str, object]:
    return {field: row[field] for field in META_FIELDS if field in row}


def content_row(row: dict[str, object]) -> dict[str, object]:
    return {field: row[field] for field in CONTENT_FIELDS if field in row}


def content_sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _base_row(path: str, *, kind: str, mime: str, size_bytes: int, is_text: int, sha256: str | None = None) -> dict[str, object]:
    value = normalize_path(path)
    row: dict[str, object] = {
        "id": path_id(value),
        "path": value,
        "basename": basename(value),
        "kind": kind,
        "ext": "" if kind == "dir" else extension(value),
        "mime": mime,
        "size_bytes": size_bytes,
        "is_text": is_text,
    }
    parent = parent_path(value)
    if parent is not None:
        row["parent"] = parent
    if sha256 is not None:
        row["sha256"] = sha256
    return row


def _with_source_metadata(row: dict[str, object], *, source_mtime_ns: int | None = None, source_size_bytes: int | None = None) -> dict[str, object]:
    if source_mtime_ns is not None:
        row["source_mtime_ns"] = int(source_mtime_ns)
    if source_size_bytes is not None:
        row["source_size_bytes"] = int(source_size_bytes)
    return row


def directory_row(path: str, *, source_mtime_ns: int | None = None) -> dict[str, object]:
    row = _base_row(
        path,
        kind="dir",
        mime="inode/directory",
        size_bytes=0,
        is_text=0,
    )
    return _with_source_metadata(row, source_mtime_ns=source_mtime_ns, source_size_bytes=0)


def infer_mime(path: str, *, fallback: str) -> str:
    guessed, _ = mimetypes.guess_type(path)
    return guessed or fallback


def text_row(
    path: str,
    text: str,
    *,
    mime: str | None = None,
    source_mtime_ns: int | None = None,
    source_size_bytes: int | None = None,
) -> dict[str, object]:
    data = text.encode("utf-8")
    row = _base_row(
        path,
        kind="file",
        mime=mime or infer_mime(path, fallback="text/plain"),
        size_bytes=len(data),
        is_text=1,
        sha256=content_sha256(data),
    )
    row["text"] = text
    return _with_source_metadata(
        row,
        source_mtime_ns=source_mtime_ns,
        source_size_bytes=len(data) if source_size_bytes is None else source_size_bytes,
    )


def bytes_row(
    path: str,
    data: bytes,
    *,
    mime: str | None = None,
    source_mtime_ns: int | None = None,
    source_size_bytes: int | None = None,
) -> dict[str, object]:
    row = _base_row(
        path,
        kind="file",
        mime=mime or infer_mime(path, fallback="application/octet-stream"),
        size_bytes=len(data),
        is_text=0,
        sha256=content_sha256(data),
    )
    row["blob_b64"] = base64.b64encode(data).decode("ascii")
    return _with_source_metadata(
        row,
        source_mtime_ns=source_mtime_ns,
        source_size_bytes=len(data) if source_size_bytes is None else source_size_bytes,
    )


def parent_directory_rows(path: str) -> list[dict[str, object]]:
    return [directory_row(value) for value in ancestor_paths(path, include_self=False)]


def target_directory_rows(path: str) -> list[dict[str, object]]:
    return [directory_row(value) for value in ancestor_paths(path, include_self=True)]


def is_probably_text(path: str, data: bytes) -> bool:
    suffix = extension(path).lower()
    if b"\x00" in data:
        return False
    if suffix in TEXT_EXTENSIONS:
        return True
    mime = infer_mime(path, fallback="application/octet-stream")
    if mime.startswith("text/"):
        return True
    if mime in {
        "application/json",
        "application/sql",
        "application/xml",
        "application/x-sh",
        "image/svg+xml",
    }:
        return True
    try:
        data.decode("utf-8")
    except UnicodeDecodeError:
        return False
    return True


def row_from_bytes(
    path: str,
    data: bytes,
    *,
    mime: str | None = None,
    source_mtime_ns: int | None = None,
    source_size_bytes: int | None = None,
) -> dict[str, object]:
    if is_probably_text(path, data):
        try:
            text = data.decode("utf-8")
        except UnicodeDecodeError:
            return bytes_row(
                path,
                data,
                mime=mime,
                source_mtime_ns=source_mtime_ns,
                source_size_bytes=source_size_bytes,
            )
        return text_row(
            path,
            text,
            mime=mime,
            source_mtime_ns=source_mtime_ns,
            source_size_bytes=source_size_bytes,
        )
    return bytes_row(
        path,
        data,
        mime=mime,
        source_mtime_ns=source_mtime_ns,
        source_size_bytes=source_size_bytes,
    )


def upsert_rows_payload(rows: Iterable[dict[str, object]]) -> dict[str, object]:
    return {
        "upsert_rows": list(rows),
        "schema": fs_schema(),
    }
